Add Valid From field

vanogrid opened the pull request:
#1032

* Adds a "Valid from" field in the Edit Card view, under Options
* Adds a "Valid from" text in the LoyaltyCardViewActivity's Info screen. The text becomes red if the date is in the future.
* Adds an icon and the "Valid from" date in the MainActivity for each card which has the date set. I couldn't find a suitable icon in Android Studio, so I tried to add the "play" symbol to the Expiry icon as you suggested. I used the "Schedule" (same as the Expiry icon) and "Play Arrow" icons from [Google Fonts](https://fonts.google.com/icons). They're open-source. Let me know how you like it!

I'm not sure if I should add some kind of validation check in the Edit Card view to make sure the "Valid from" date is before the "Expiry" date. Do you think it's needed?

vanogrid commented on app/src/main/res/values/strings.xml:
```diff
@@ -300,4 +300,8 @@
     <string name="welcome">Welcome to Catima</string>
+    <string name="validFromDate">Valid from</string>
+    <string name="anyDate">Any date</string>
+    <string name="chooseValidFromDate">Choose Valid from date</string>
```
I'm not sure if I've capitalized "Valid from" here correctly, as English is not my native language. Please have a look. The string for the expiry date is "Choose expiry date"

TheLastProject reviewed (changes requested):
> I'm not sure if I should add some kind of validation check in the Edit Card view to make sure the "Valid from" date is before the "Expiry" date. Do you think it's needed?

I think we should look into if we can specify a minimum and maximum date on the calendar. Right now it's not a problem if expiry is before start, but I think forcing the values to be correct will help us when adding more features in the future.

---

The review I did was based on just reading code as I haven't had the time to run this yet (Hacktoberfest has caused a lot of contributions, which is awesome but also taking a lot of my time, so I can't be as thorough as I normally like to be without delaying contributors too much)

TheLastProject commented on app/src/main/java/protect/card_locker/LoyaltyCardCursorAdapter.java:
```diff
@@ -308,6 +315,33 @@ public void setBalanceField(BigDecimal balance, Currency balanceType) {
             mBalanceField.requestLayout();
         }
 
+        public void setValidFromField(Date validFrom) {
+            if (validFrom == null) {
+                mValidFromField.setVisibility(View.GONE);
+            } else {
+                int size = mSettings.getFontSizeMax(mSettings.getSmallFont());
+                int drawableSize = dpToPx((size * 24) / 14, mContext);
+                mDivider.setVisibility(View.VISIBLE);
+                mValidFromField.setVisibility(View.VISIBLE);
+                Drawable validFromIcon = mValidFromField.getCompoundDrawables()[0];
+                if (validFromIcon != null) {
+                    validFromIcon.setBounds(0, 0, drawableSize, drawableSize);
+                    mValidFromField.setCompoundDrawablesRelative(validFromIcon, null, null, null);
+                    if (Utils.isNotYetValid(validFrom)) {
+                        validFromIcon.setColorFilter(BlendModeColorFilterCompat.createBlendModeColorFilterCompat(Color.RED, BlendModeCompat.SRC_ATOP));
+                    } else if (mDarkModeEnabled) {
+                        validFromIcon.setColorFilter(BlendModeColorFilterCompat.createBlendModeColorFilterCompat(Color.WHITE, BlendModeCompat.SRC_ATOP));
+                    }
+                }
+                mValidFromField.setText(DateFormat.getDateInstance(DateFormat.LONG).format(validFrom));
+                if (Utils.isNotYetValid(validFrom)) {
+                    mValidFromField.setTextColor(Color.RED);
+                }
+                mValidFromField.setTextSize(size);
+            }
+            mValidFromField.requestLayout();
+        }
+
```
I just merged #1088 which should make it so you don't have to copy-paste all of this and can just call one function. Please pull in the latest changes from master (rebase preferred but everything is good) and use that :)

vanogrid replied:
Thanks! I've rebased my branch :)

vanogrid commented on app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:
```diff
@@ -366,6 +369,39 @@ public void onTextChanged(CharSequence s, int start, int before, int count) {
+                validFromList.add(1, getString(R.string.chooseValidFromDate));
+                ArrayAdapter<String> validFromAdapter = new ArrayAdapter<>(LoyaltyCardEditActivity.this, android.R.layout.select_dialog_item, validFromList);
+                validFromField.setAdapter(validFromAdapter);
+            }
```
done

vanogrid commented on app/src/main/java/protect/card_locker/LoyaltyCardViewActivity.java:
```diff
@@ -510,7 +524,7 @@ private void showInfoDialog() {
 
                 infoText.append(getString(R.string.expiryStateSentenceExpired, formattedExpiry));
-                infoText.setSpan(new ForegroundColorSpan(Color.RED), start, infoText.length(), Spanned.SPAN_INCLUSIVE_INCLUSIVE);
+                infoText.setSpan(new ForegroundColorSpan(Color.RED), start, infoText.length(), Spanned.SPAN_INCLUSIVE_EXCLUSIVE);
```
I noticed a bug with the text where if the 'valid from' text was red, the 'expiry' text was also red even if it shouldn't have to be. I assume using `infoText.length()` as the `end` value was causing the issue, but I didn't debug it enough to find the actual cause. I think using `SPAN_INCLUSIVE_EXCLUSIVE` for both 'valid from' and 'expiry' should prevent this bug from appearing again if more text is added to the InfoDialog after the 'expiry' date.

vanogrid commented:
At first I didn't extract the duplicated logic between 'valid from' and 'expiry' into common functions in case in the future some new functionality needed to be added to only one of them, but it makes sense to reduce the duplication now and split the logic later if needed :)

I added a new commit for specifying min and max values for the DatePicker. I didn't merge it with the rest of the changes, so it's easier for review. Let me know what you think.

vanogrid commented on app/src/main/java/protect/card_locker/LoyaltyCardEditActivity.java:
```diff
@@ -1282,17 +1295,23 @@ public Dialog onCreateDialog(Bundle savedInstanceState) {
             DatePickerDialog datePickerDialog = new DatePickerDialog(getActivity(), this, year, month, day);
-            datePickerDialog.getDatePicker().setMinDate(getMinDateOfDatePicker());
+            datePickerDialog.getDatePicker().setMinDate(minDate != null ? minDate.getTime() : getDefaultMinDateOfDatePicker());
+            datePickerDialog.getDatePicker().setMaxDate(maxDate != null ? maxDate.getTime() : getDefaultMaxDateOfDatePicker());
```
Technically I don't need to set the default max date here, but having an if statement here didn't look that neat to me 😄 :
```
if (maxDate != null) {
    datePickerDialog.getDatePicker().setMaxDate(maxDate.getTime());
}
```
Let me know if I should use the if statement instead :)

TheLastProject reviewed (changes requested):
Sorry, merging something else caused some conflicts, would be great if you could resolve the conflict.

I'm not really seeing major issues with the code now, I'll re-read soon (probably tomorrow) to make sure. I'm not completely sure about merging this though because having an extra field will just make #1049 more noticeable so I do feel I'd rather merge that first, but if I see no issues my next review I'll at least label it hacktoberfest-accepted so it counts as it should

vanogrid commented on app/src/main/java/protect/card_locker/Utils.java:
```diff
@@ -208,16 +208,23 @@ static private BarcodeValues getBarcodeFromBitmapReal(Bitmap bitmap) {
     }
 
+    static public Boolean isNotYetValid(Date validFromDate) {
+        return validFromDate.after(getStartOfToday().getTime());
```
It's currently working the way you expect it to. The `validFromDate` is stored at midnight in DatePickerFragment's `onDateSet` method (October 16 00:00:00.000) and `getStartOfToday` also returns the today's date at midnight, so on October 16 `after` will return false, i.e. the date is valid.

I realize I can merge the two methods `isNotYetValid` and `hasExpired` into a single `isValid` method, so the code is easier to maintain. What do you think?

TheLastProject replied:
> It's currently working the way you expect it to. The validFromDate is stored at midnight in DatePickerFragment's onDateSet method (October 16 00:00:00.000) and getStartOfToday also returns the today's date at midnight, so on October 16 after will return false, i.e. the date is valid.

That's not what's happening for me on your branch. If I right now on October 17th create a card that's Valid From October 17th it is marked as red for me:

![image](https://user-images.githubusercontent.com/1885159/196243489-aec8cb93-f13a-4ee6-8153-210c00a4438b.png)

> I realize I can merge the two methods isNotYetValid and hasExpired into a single isValid method, so the code is easier to maintain. What do you think?

I think I'd prefer a name like `isTodayOrLater`, it's a bit clearer than `isValid`.

vanogrid replied:
Sorry, you're right. I'll try to fix it asap.

vanogrid replied:
I couldn't make it work with a single method like I imagined it 😞 At least I managed to fix the bug.

The `Calendar` instance in `Utils.getStartOfToday` is instantiated as `new GregorianCalendar()` and the one in `DatePickerFragment.onDateSet` is instantiated with `Calendar.getInstance()`. It turns out the default value for `Calendar.AM_PM` is different between them (could be because of the phone's regional settings) and the first one sets the hour at midnight (12:00 AM) while the other sets it at noon (12:00 PM). 

Theoretically, adding something like `c.set(Calendar.AM_PM, Calendar.AM);` to the DatePickerFragment's Calendar could have worked but only for new cards, as the existing cards are already saved in the DB at noon. I'm not too keen on trying not to break the DB with a migration script 😄 Instead, I added a check if the date is not today.

TheLastProject replied:
Oh geez, that's a subtle bug, nice catch!

I wonder though, with how you're explaining this, wouldn't that mean that the expiryCheck is also broken (firing 12 hours too early or too late)?

I'm kinda worried now about what else may be wrong, I wonder what other calendar types exist and what other subtle bugs might've creeped in with the `Calendar.getInstance()` and `new GregorianCalendar()` difference :fearful:

TheLastProject replied:
That's great research!

So, if I understand things correctly some small fixes there would be:
1. Maybe change the current `Calendar.getInstance()` to `new GregorianCalendar()` just to be safe and prevent possible bugs due to possible changes in the future (because right now it's not synced up and I don't have any clue how bad breakage may be if Android adds a new calendar in the future and we start comparing between calendar types).
2. Change the code in `DatePickerFragment.onDateSet()` to use `HOUR_OF_DAY` instead of `HOUR` so new database entries are written correctly.
3. Add some note in the `hasExpired()` explaining the previous bug and how it could be 12 hours late for cards saved before the `HOUR_OF_DAY` fix.

It won't fix the weirdness for existing cards, but it'll gradually transition every single user out of this weird little bug I believe.

I would say these fixes would fit in a new PR as they're not directly part of this new feature?

vanogrid replied:
Yes, I believe the steps you described will fix the bug for the newly created cards. I'll create a PR.

vanogrid commented:
Rebased and resolved the merge conflicts. 

It's perfectly ok to wait for #1049. I also noticed it when testing. Just ping me if new merge conflicts need to be resolved before merging :)

vanogrid commented:
I rebased the branch after #1103 was merged. I also removed the `!DateUtils.isToday()` check in `Utils.isNotYetValid()` since the date is now always being stored at 12:00 AM and the check is not needed.

TheLastProject commented:
Thanks again :)
